packages/waku/src/lib/renderers/html.ts

```diff
@@ -0,0 +1,269 @@
+      }
+      pending.push(chunk);
+      if (/<\/\w+>$/.test(decoder.decode(chunk))) {
+        clearTimeout(timer);
+        timer = setTimeout(() => {
+          controller.enqueue(concatUint8Arrays(pending.splice(0)));
+        });
+      }
+    },
+    flush(controller) {
+      clearTimeout(timer);
+      if (pending.length) {
+        controller.enqueue(concatUint8Arrays(pending.splice(0)));
+      }
+    },
+  });
+};
+
+export function renderHtml(
+  config: PureConfig,
+  ctx: HandlerContext,
+  htmlHead: string,
+  elements: Elements,
+  html: ReactNode,
+  rscPath: string,
+): ReadableStream {
+  const modules = ctx.unstable_modules;
+  if (!modules) {
+    throw new Error('handler middleware required (missing modules)');
+  }
+  const {
+    default: { renderToReadableStream },
+  } = modules.rdServer as { default: typeof RDServerType };
+  const {
+    default: { createFromReadableStream },
+  } = modules.rsdwClient as { default: typeof RSDWClientType };
+  const { ServerRoot } =
+    modules.wakuMinimalClient as typeof WakuMinimalClientType;
+
+  const stream = renderRsc(config, ctx, elements);
+  const htmlStream = renderRscElement(config, ctx, html);
+  const isDev = !!ctx.unstable_devServer;
+  const rootDir = ctx.unstable_devServer?.rootDir || '';
+  const moduleMap = new Proxy(
+    {} as Record<string, Record<string, ImportManifestEntry>>,
+    {
+      get(_target, filePath: string) {
+        return new Proxy(
+          {},
+          {
+            get(_target, name: string) {
+              if (isDev) {
+                // TODO too long, we need to refactor this logic
+                let file = filePath
+                  .slice(config.basePath.length)
+                  .split('?')[0]!;
+                const isFsPath = file.startsWith('@fs/');
+                file = isFsPath ? file.slice('@fs'.length) : file;
+                const fileWithAbsolutePath = isFsPath
+                  ? file
+                  : encodeFilePathToAbsolute(joinPath(rootDir, file));
+                const wakuDist = joinPath(
+                  fileURLToFilePath(import.meta.url),
+                  '../../..',
+                );
+                if (fileWithAbsolutePath.startsWith(wakuDist)) {
+                  const id =
+                    'waku' +
+                    fileWithAbsolutePath
+                      .slice(wakuDist.length)
+                      .replace(/\.\w+$/, '');
+                  (globalThis as any).__WAKU_CLIENT_CHUNK_LOAD__(id);
+                  return { id, chunks: [id], name };
+                }
+                const id = filePathToFileURL(file);
+                (globalThis as any).__WAKU_CLIENT_CHUNK_LOAD__(id);
+                return { id, chunks: [id], name };
+              }
+              // !isDev
+              const id = filePath.slice(config.basePath.length);
+              (globalThis as any).__WAKU_CLIENT_CHUNK_LOAD__(id);
+              return { id, chunks: [id], name };
```

Comment:
Might be more readable if this is put at the top of the function as an early return when `!isDev`

Reply:
on second thought, it's the problem of `if (isDev)` is too long. Let's leave it for now.